### PR TITLE
appengine/gophers/6: first version of task queues

### PR DESCRIPTION
**Pull request by tbpg**
This requires configuration of a bucket name. Is there a good default to use? Another option is to reuse the Firebase bucket (project-id.appspot.com).

cc @angelafunk

### CODE REVIEW

**Comment by broady** [comment]
Re-using the bucket SGTM. This is the default bucket for the GAE app.

File: appengine/gophers/gophers-6/main.go [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+
+// labelFunc will be called asynchronously as a Cloud Task. labelFunc can
+// be executed by calling labelFunc.Call(ctx, postID). If an error is returned
+// the function will be retried.
+var labelFunc = delay.Func("label-image", func(ctx context.Context, id int64) error {
+	// Get the post to label.
+	k := datastore.NewKey(ctx, "Post", "", id, nil)
+	post := Post{}
+	err := datastore.Get(ctx, k, &post)
+	if err != nil {
```
Comment (broady) [inline appengine/gophers/gophers-6/main.go]: prefer compacted style:

`if err := ...; err != nil {`

File: appengine/gophers/gophers-6/main.go [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+		Posted:  time.Now(),
+	}
+	params.Name = post.Author
+
+	// Get the image if there is one.
+	imageURL, err := uploadFileFromForm(ctx, r)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		params.Notice = "Error saving image: " + err.Error()
+		indexTemplate.Execute(w, params)
```
Comment (broady) [inline appengine/gophers/gophers-6/main.go]: I guess we should retain their message, too, not just their name?

**Comment by broady** [comment]
LGTM

File: appengine/gophers/gophers-6/main.go (hunk 1 above)
Comment (tbpg) [reply appengine/gophers/gophers-6/main.go]: Done.

File: appengine/gophers/gophers-6/main.go [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+	indexTemplate = template.Must(template.ParseFiles("index.html"))
+)
+
+type Post struct {
+	Author   string
+	UserID   string
+	Message  string
+	Posted   time.Time
+	ImageURL string
+	Labels   string
```
[the remark this answers is not in the document]
Comment (tbpg) [reply appengine/gophers/gophers-6/main.go]: Done.

File: appengine/gophers/gophers-6/main.go (hunk 2 above)
Comment (tbpg) [reply appengine/gophers/gophers-6/main.go]: Done.

File: appengine/gophers/gophers-6/main.go [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+	w.ACL = []storage.ACLRule{{Entity: storage.AllUsers, Role: storage.RoleReader}}
+	w.ContentType = fh.Header.Get("Content-Type")
+
+	// Entries are immutable, be aggressive about caching (1 day).
+	w.CacheControl = "public, max-age=86400"
+
+	if _, err := io.Copy(w, f); err != nil {
+		return "", err
+	}
+	if err := w.Close(); err != nil {
```
Comment (broady) [inline appengine/gophers/gophers-6/main.go]: CloseWithError (from the io.Copy)

**Pull request merged**